### PR TITLE
Suprsync timecode dir finalization

### LINKED ISSUE #444: Suprsync needs to pass completion information to g3tsmurf and imprinter
As discussed on the data flow call. We need to know "when all data have been transferred" from the smurf server to level 2 and then  we'll be building g3tsmurf and imprinter completion off of this one. 

Idea: 

- Write “finalized until XXXXX” file to smurf / timestream folders. This will be one file per suprsync agent ( per folder(smurf/timestream) & per server ). 
- G3tsmurf will just use the earliest “complete” time in the top level folder. Will need structure changes to g3tsmurf to track this nicely.

### PR DESCRIPTION
**Pull request by jlashner**
Adds timecode directory tracking and finalization to suprsync.

## Description

This PR adds a new table to the suprsync db that tracks "timecode directories" or directories where the first subdirectory (in the remote path) is a 5-digit timecode. The agent will automatically determine whether a timecode directory is "completed" (when there are no more expected files) and synced, and will then write and sync a "finalization file" that can be used by upstream data packaging software to determine if a timecode directory is ready to be bookbound.

## Motivation and Context

Required for proper bookbinding of `smurf` books.

Resolves #444.

## How Has This Been Tested?

Tests ran locally. We may want to test the agent on the e2e system.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [X] My code follows the code style of this project.
- [X] My change requires a change to the documentation.
- [X] I have updated the documentation accordingly.

### CODE REVIEW

File: socs/db/suprsync.py [hunk 1]
```diff
@@ -278,6 +334,116 @@ def get_known_files(self, archive_name, session=None):
...
+        Takes the next series of actions for a timecode dir object.
+        - If we expect no more files to be added to the tc dir, marks it as
+          complete
+        - If all files in the tc dir have been synced, marks it as synced
+        - If the tc dir is synced and not finalized, creates the finalization
+          file and marks as finalized.
+        """
+        if tcdir.finalized:
+            return
+
```
Comment (mhasself) [inline socs/db/suprsync.py]: You call time.time() a bunch of times below... I would be more comfortable if you computed "now = time.time()" and then used that everywhere... fewer race conditions / consistency concerns.

File: socs/db/suprsync.py [hunk 2]
```diff
@@ -278,6 +334,116 @@ def get_known_files(self, archive_name, session=None):
...
+            for tc, in all_tcs:
+                if tc > tcdir.timecode:
+                    # Mark as complete if there's a timecode after this one
+                    tcdir.completed = True
+                    break
+            else:
+                # No timecodes after this one. Mark after complete if we are
+                # over a full day away.
+                if (time.time()//1e5 - tcdir.timecode) > 1:
+                    tcdir.completed = True
```
Comment (mhasself) [inline socs/db/suprsync.py]: Please make this auto-completion time a constant at the top of the program.

File: docs/agents/suprsync.rst [hunk 3]
```diff
@@ -170,6 +172,62 @@ timestream files (using the archive name "timestreams"), as described in
...
+that will contain information about the timecode directory, including the
+number of files that this suprsync instance has synced over, the sub-directories
+that this suprsync instance has added to, the instance-id of the suprsync agent,
+and the finalization time. The directory on the local host that these files will
+be written to is set using the ``--suprsync-file-root`` argument, and the
+path on the remote host will be automatically generated.
+
+Timecode Dir Completion Requirements
+```````````````````````````````````
+A timecode directory will be marked as complete if:
```
Comment (mhasself) [inline docs/agents/suprsync.rst:209]: "+ any of the following"

**Review by mhasself — APPROVED** [review]
Thanks!  Looks right to me.

**Review by kmharrington — APPROVED** [review]
Looks good to me!

File: socs/db/suprsync.py [hunk 4]
```diff
@@ -17,6 +19,40 @@
 Base = declarative_base()
 
 
+class TimecodeDir(Base):
```
Comment (BrianJKoopman) [inline socs/db/suprsync.py:22]: This docstring is great, please add an autoclass to the docs page under "Supporting APIs".

File: socs/db/suprsync.py [hunk 5]
```diff
@@ -4,9 +4,11 @@
 import time
 
 import txaio
-from sqlalchemy import Boolean, Column, Float, Integer, String, create_engine
+import yaml
+from sqlalchemy import (Boolean, Column, Float, ForeignKey, Integer, String,
+                        create_engine)
 from sqlalchemy.ext.declarative import declarative_base
-from sqlalchemy.orm import sessionmaker
+from sqlalchemy.orm import relationship, sessionmaker
```
Comment (BrianJKoopman) [inline socs/db/suprsync.py:11]: `sqlalchemy.orm.relationship` is an unused import, remove.

File: tests/agents/test_suprsync_agent.py [hunk 6]
```diff
@@ -1,9 +1,11 @@
 import os
+import time
 
 import numpy as np
 import txaio
 
-from socs.db.suprsync import SupRsyncFileHandler, SupRsyncFilesManager
+from socs.db.suprsync import (SupRsyncFile, SupRsyncFileHandler,
```
Comment (BrianJKoopman) [inline tests/agents/test_suprsync_agent.py:7]: `SupRsyncFile` is an unused import, remove.

File: socs/db/suprsync.py [hunk 7]
```diff
@@ -76,6 +112,26 @@ def __str__(self):
...
 
 
+def split_path(path):
+    """Splits path into a list where each element is a subdirectory"""
+    return os.path.normpath(path).strip('/').split('/')
+
+
+def check_timecode(file: SupRsyncFile):
+    """
+    Tries to extract timecode from the remote path. If it failes, returns
```
Comment (BrianJKoopman) [inline socs/db/suprsync.py:122]: 'failes' -> 'fails'

**Review by BrianJKoopman — CHANGES_REQUESTED** [review]
Looks good! Two unused imports to remove and two small docs requests.

**Review by BrianJKoopman — CHANGES_REQUESTED** [review]
I pushed a small docs change.

Can you look into why [this test](https://github.com/simonsobs/socs/actions/runs/5246750172/jobs/9475983754?pr=454) is failing?

We need to resolve that before merging.

**Comment by jlashner** [comment]
I merged the SupRsync CLI branch into this, and fixed the failing tests, and looks like everythings working now. Do you think we can merge this?

**Review by BrianJKoopman — COMMENTED** [review]
> I merged the SupRsync CLI branch into this, and fixed the failing tests, and looks like everythings working now. Do you think we can merge this?

I think there's still one unresolved comment from @mhasself.

File: socs/db/suprsync.py (hunk 2 above)
Comment (jlashner) [reply socs/db/suprsync.py]: This has been resolved but it's not showing it in the textbox, though it's been marked as outdated. If you look at the "files changed" I did indeed replace all the `time.time()` calls with a constant.

Comment (jlashner) [reply socs/db/suprsync.py]: unless I'm misinterpreting the feedback

Comment (jlashner) [reply socs/db/suprsync.py]: Talked with brian about this in slack and decided to move the 1 here to a constant defined at the top of the file. I think I misinterpreted @mhasself's initial comment.

**Review by BrianJKoopman — APPROVED** [review]

**Pull request merged**